python/runtime/cudaq/algorithms/py_state.cpp

```diff
@@ -252,6 +252,74 @@ static py::buffer_info getCupyBufferInfo(py::buffer cupy_buffer) {
   );
 }
 
+static cudaq::state createStateFromPyBuffer(py::buffer data,
+                                            LinkedLibraryHolder &holder) {
+  const bool isHostData = !py::hasattr(data, "__cuda_array_interface__");
+  // Check that the target is GPU-based, i.e., can handle device
+  // pointer.
+  if (!holder.getTarget().config.GpuRequired && !isHostData)
+    throw std::runtime_error(
+        fmt::format("Current target '{}' does not support CuPy arrays.",
+                    holder.getTarget().name));
+
+  auto info = isHostData ? data.request() : getCupyBufferInfo(data);
+  if (info.shape.size() > 2)
+    throw std::runtime_error(
+        "state.from_data only supports 1D or 2D array data.");
+  if (info.format != py::format_descriptor<std::complex<float>>::format() &&
+      info.format != py::format_descriptor<std::complex<double>>::format())
+    throw std::runtime_error(
+        "A numpy array with only floating point elements passed "
+        "to "
+        "state.from_data. input must be of complex float type, "
+        "please "
+        "add to your array creation `dtype=numpy.complex64` if "
+        "simulation is FP32 and `dtype=numpy.complex128` if "
+        "simulation if FP64, or dtype=cudaq.complex() for "
+        "precision-agnostic code");
+
+  if (!isHostData || info.shape.size() == 1) {
+    if (info.format == py::format_descriptor<std::complex<float>>::format()) {
+      return state::from_data(std::make_pair(
+          reinterpret_cast<std::complex<float> *>(info.ptr), info.size));
+    } else {
```

Comment:
nit: style. else after return and braces.

Reply:
Fixed.